commons/src/build_output.rs

```diff
@@ -0,0 +1,647 @@
+use crate::fun_run::{self, CmdError};
+use libherokubuildpack::command::CommandExt;
+use libherokubuildpack::write::line_mapped;
+use libherokubuildpack::write::mappers::add_prefix;
+use std::io::Write;
+use std::{
+    process::{Command, Output},
+    time::{Duration, Instant},
+};
+
+mod time {
+    use std::time::Duration;
+
+    // Returns the part of a duration only in miliseconds
+    pub(crate) fn milliseconds(duration: &Duration) -> u32 {
+        duration.subsec_millis()
+    }
+
+    pub(crate) fn seconds(duration: &Duration) -> u64 {
+        duration.as_secs() % 60
+    }
+
+    pub(crate) fn minutes(duration: &Duration) -> u64 {
+        (duration.as_secs() / 60) % 60
+    }
+
+    pub(crate) fn hours(duration: &Duration) -> u64 {
+        (duration.as_secs() / 3600) % 60
+    }
+
+    pub(crate) fn human(duration: &Duration) -> String {
+        let hours = hours(duration);
+        let minutes = minutes(duration);
+        let seconds = seconds(duration);
+        let miliseconds = milliseconds(duration);
+
+        if hours > 0 {
+            format!("{hours}h {minutes}m {seconds}s")
+        } else if minutes > 0 {
+            format!("{minutes}m {seconds}s")
+        } else if seconds > 0 || miliseconds > 100 {
+            // 0.1
+            format!("{seconds}.{miliseconds:0>3}s")
+        } else {
+            String::from("< 0.1s")
+        }
+    }
+
+    #[cfg(test)]
+    mod test {
+        use super::*;
+
+        #[test]
+        fn test_millis_and_seconds() {
+            let duration = Duration::from_millis(1024);
+            assert_eq!(24, milliseconds(&duration));
+            assert_eq!(1, seconds(&duration));
+        }
+
+        #[test]
+        fn test_display_duration() {
+            let duration = Duration::from_millis(99);
+            assert_eq!("< 0.1s", human(&duration).as_str());
+
+            let duration = Duration::from_millis(1024);
+            assert_eq!("1.024s", human(&duration).as_str());
+
+            let duration = std::time::Duration::from_millis(60 * 1024);
+            assert_eq!("1m 1s", human(&duration).as_str());
+
+            let duration = std::time::Duration::from_millis(3600 * 1024);
+            assert_eq!("1h 1m 26s", human(&duration).as_str());
+        }
+    }
+}
+
+// Helper for printing without newlines that auto-flushes stdout
+fn raw_inline_print(contents: impl AsRef<str>) {
+    let contents = contents.as_ref();
+    print!("{contents}");
+    std::io::stdout().flush().expect("Stdout is writable");
+}
+
+/// All work is done inside of a section. Advertize a section topic
+pub fn section(topic: impl AsRef<str>) -> section::Section {
+    let topic = String::from(topic.as_ref());
+    println!("- {topic}");
+
+    section::Section { topic }
+}
+
+/// Top level buildpack header
+///
+/// Should only use one per buildpack
+#[must_use]
+pub fn header(buildpack: impl AsRef<str>) -> BuildpackDuration {
```

Comment:
I did want something more specific and like that. 

I've been thinking about how this will apply to buildpacks and, specifically, multi-buildpacks. I.e. when the Node buildpack executes, do we want it to advertise each-and-every engine? One UX issue with he current output from Ruby's perspective is that there's no indication why the node engine is running in the body of the build output. One thought I had was if we could communicate with the dependency somehow and get some output like:

```
# Heroku Node Engine (for `heroku/ruby`)
...
```